botfront/imports/api/graphql/project/import.utils.js

```diff
@@ -0,0 +1,183 @@
+
+import { determineDataType } from '../../../lib/importers/common';
+import {
+    loadBotfrontConfig, loadRasaConfig, loadConversations, loadIncoming, loadEndpoints, loadCredentials, doValidation,
+} from '../../../lib/importers/loadMisc.js';
+import {
+    loadDomain,
+} from '../../../lib/importers/loadDomain.js';
+import { handleImportAll } from './fileImporters';
+import Project from './project.model';
+
+function streamToString (stream) {
+    const chunks = [];
+    return new Promise((resolve, reject) => {
+        stream.on('data', chunk => chunks.push(chunk));
+        stream.on('error', reject);
+        stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
+    });
+}
+
+
+// generate a array of info messages that says what will be imported
+// the number of stories, of storiy-groups etc
+export function generateSummary(files) {
+    // TODO
+    return null;
+}
+// dispatch file to correct loader and check for errors
+export function loadFile(file, params) {
+    const { dataType } = file;
+    if (dataType === 'domain') {
+        return loadDomain({ file, params });
+    }
+    // if (dataType === 'stories') {
+    //     return addStoryFile({ file, ...params });
+    // }
+    // if (dataType === 'nlu') {
+    //     return addNluFile({ file, ...params });
+    // }
+    if (dataType === 'conversations') {
+        return loadConversations({ file, params });
+    }
+    if (dataType === 'incoming') {
+        return loadIncoming({ file, params });
+    }
+    if (dataType === 'bfconfig') {
+        return loadBotfrontConfig({ file, params });
+    }
+    if (dataType === 'rasaconfig') {
+        return loadRasaConfig({ file, params });
+    }
+    if (dataType === 'endpoints') {
+        return loadEndpoints({ file, params });
+    }
+    if (dataType === 'credentials') {
+        return loadCredentials({ file, params });
+    }
+    return { file, errors: ['unknown file format'] };
+}
+
+// extract the raw text from the files and infer types
+// if there is a bfconfig file it process it, because we need the data from that file for the validation later
+// (eg: the default domain, the project languages)
+export async function getRawTextAndType(files, params) {
+    const projectConfig = null;
+    const filesDataAndTypes = await Promise.all(files.map(async (file) => {
+        const { filename } = file;
+        if (file.filename.match(/\.(yml|json|md)/)) {
+            const rawText = await streamToString(file.createReadStream());
+            if (/\ufffd/.test(rawText)) { // out of range char test
+                return {
+                    file,
+                    filename,
+                    errors: ['file is not parseable text'],
+                };
+            }
+            
+            const dataType = determineDataType(file, rawText);
+            if (dataType === 'bfconfig' && projectConfig === null) { // that way we take the first file if there are multiple project files
```

Comment:
but `projectConfig` is hardcoded? I think we should have a more general merge mechanism; refer to thoughts about "bfconfig" in comments